### ISSUE TITLE
test_cpp_extensions_aot_* fail on Power

### ISSUE THREAD

**Issue by Flamefire** [pytorch/pytorch#41261]
## 🐛 Bug

`python run_test.py -i test_cpp_extensions_aot_ninja` and the no_ninja one fail with a lot of compiler errors

## To Reproduce

1. Build and install on power
1.`python run_test.py -i test_cpp_extensions_aot_ninja`

The first error:
```
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/c10/macros/Macros.h:146:47: Fehler: »__vector(4) __bool int« kann nicht nach »long int« umgewandelt werden
 #define C10_UNLIKELY(expr)  (__builtin_expect(static_cast<bool>(expr), 0))
                                               ^~~~~~~~~~~~~~~~~~~~~~~
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/c10/util/Exception.h:219:39: Anmerkung: in Definition des Makros »C10_EXPAND_MSVC_WORKAROUND«
 #define C10_EXPAND_MSVC_WORKAROUND(x) x
                                       ^
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/c10/util/Exception.h:238:34: Anmerkung: bei Substitution des Makros »C10_UNLIKELY«
 #define C10_UNLIKELY_OR_CONST(e) C10_UNLIKELY(e)
                                  ^~~~~~~~~~~~
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/c10/util/Exception.h:285:7: Anmerkung: bei Substitution des Makros »C10_UNLIKELY_OR_CONST«
   if (C10_UNLIKELY_OR_CONST(!(cond))) {       \
       ^~~~~~~~~~~~~~~~~~~~~
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/c10/util/Exception.h:447:32: Anmerkung: bei Substitution des Makros »TORCH_INTERNAL_ASSERT«
     C10_EXPAND_MSVC_WORKAROUND(TORCH_INTERNAL_ASSERT(cond, __VA_ARGS__));     \
                                ^~~~~~~~~~~~~~~~~~~~~
/lustre/ssd/ws/s3248973-EasyBuild/easybuild-ml/software/PyTorch/1.6.0-fosscuda-2019b-Python-3.7.4/lib/python3.7/site-packages/torch/include/ATen/native/DispatchStub.h:78:7: Anmerkung: bei Substitution des Makros »AT_ASSERTM«
       AT_ASSERTM(cuda_dispatch_ptr, "DispatchStub: missing CUDA kernel");
       ^~~~~~~~~~
<eingebaut>: Anmerkung:   Argument 1 von »long int __builtin_expect(long int, long int)« wird initialisiert
```

There are many errors, some are similar some are different. The same setup works fine on an x86 node.

## Environment

I tested PyTorch 1.5.1 and 1.6.0-rc2 on a POWER9 system with RHEL 7.5

cc @malfet @seemethere @walterddr @yf225 @glaringlee

**Comment by Flamefire** [comment]
This still applies on 1.7.0

**Comment by glaringlee** [comment]
Mark this 'triage review'. 
This should be discussed in triage meeting see what level we can support for PowerPC at this point.

**Comment by Flamefire** [comment]
Looks like this is caused by the header `altivec.h` which defines `bool` (and `vector`) to `__bool` leading to an error due to context sensitive keywors being disallowed in standard C++: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=58241

The solution is to undefine `bool` and `vector`. See also https://gitlab.mpcdf.mpg.de/bbramas/inastemp/blob/9e153cdaebb88d5dd1dece5766ce98b88a52d681/Src/ALTIVEC/InaVecALTIVECFloat.hpp#L19-22 and https://github.com/eigenteam/eigen-git-mirror/blob/36b95962756c1fce8e29b1f8bc45967f30773c00/Eigen/src/Core/util/ConfigureVectorization.h#L366-L371

**Comment by Flamefire** [comment]
Fix added in #46179 
Would be great to include this in 1.7

**Comment by glaringlee** [comment]
@Flamefire Nice fix.
I will approve this, whether we can put this into 1.7 is not decided yet, I will follow this up and update here.

**Comment by glaringlee** [comment]
@Flamefire 
Sorry about the inconvenience, this code will be included in the next release, since 1.7 release is already cut and we want this code stay for longer before in 'release' status. Please built it with master branch at this point. Thanks.

**Comment by quickwritereader** [comment]
@glaringlee 
why here, we choose undefine  instead of using -std=g++11  ?
I rebased my Pr and it started to give me errors. I think I can still do a workaround even with undefine things

**Comment by Flamefire** [comment]
@quickwritereader What do you mean? The undefine is definitely needed and pretty much the only valid fix. This header should not define C/C++ keywords to something else as that is explicitely forbidden by the standard (at least defines as "undefined behavior")

If this is causing any trouble, let's discuss that in a separate issue and find a solution.

**Comment by quickwritereader** [comment]
I meant using -std = gnu++  which consider such violations.

**Comment by Flamefire** [comment]
No it does not: It is already done by the preprocessor so the C++ lexer/parser can't do anything anymore. Test case: https://godbolt.org/z/T185zs

**Comment by quickwritereader** [comment]
actually, I meant these: 
-std=gnu++11 used with include <altivec.h>  works : https://godbolt.org/z/Kn4Eq3
-std=c++11 used with include <altivec.h>  does not work: https://godbolt.org/z/oE9f68

**Comment by Flamefire** [comment]
Ah I see. Reason is that with `-std=gnu++11` GCC defines `__APPLE_ALTIVEC__` which does skip the bool define. Found something on the GCC ML which introduced that in some version: https://gcc.gnu.org/legacy-ml/gcc-patches/2004-05/msg00454.html

Anyway: I don't see why we should force using the non-standard C++ mode. The source explicitly says `You are allowed to undef these for C++ compatibility`, so that's what I did.

If some code is having trouble with that it can use the `__bool` type instead.
